pkg/datastore/filedb/filter.go

```diff
@@ -15,19 +15,177 @@
 package filedb
 
 import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode"
+
 	"github.com/pipe-cd/pipecd/pkg/datastore"
 )
 
-// TODO: Implement filterable interface for each collection.
 type filterable interface {
 	Match(e interface{}, filters []datastore.ListFilter) (bool, error)
 }
 
 func filter(col datastore.Collection, e interface{}, filters []datastore.ListFilter) (bool, error) {
 	fcol, ok := col.(filterable)
-	if !ok {
+	if ok {
+		return fcol.Match(e, filters)
+	}
+
+	// remarshal entity as map[string]interface{} struct.
+	raw, _ := json.Marshal(e)
+	var omap map[string]interface{}
+	if err := json.Unmarshal(raw, &omap); err != nil {
+		return false, err
+	}
+
+	for _, filter := range filters {
+		field := convertCamelToSnake(filter.Field)
+		if strings.Contains(field, ".") {
+			// TODO: Handle nested field name such as SyncState.Status.
+			return false, datastore.ErrUnsupported
+		}
+
+		val, ok := omap[field]
+		// If the object does not contain given field name in filter, return false immidiately.
+		if !ok {
+			return false, nil
+		}
+
+		cmp, err := compare(val, filter.Value, filter.Operator)
+		if err != nil {
+			return false, err
+		}
+
+		if !cmp {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func compare(val, operand interface{}, op datastore.Operator) (bool, error) {
+	var valNum, operandNum int64
+	switch v := val.(type) {
+	case int, int8, int16, int32, int64:
+		valNum = reflect.ValueOf(v).Int()
+	case uint, uint8, uint16, uint32, uint64:
+		valNum = reflect.ValueOf(v).Int()
+	default:
+		if datastore.IsNumericOperator(op) {
+			return false, fmt.Errorf("value of type unsupported")
+		}
+	}
+	switch o := operand.(type) {
+	case int, int8, int16, int32, int64:
+		operandNum = reflect.ValueOf(o).Int()
+	case uint, uint8, uint16, uint32, uint64:
+		operandNum = reflect.ValueOf(o).Int()
+	default:
+		if datastore.IsNumericOperator(op) {
+			return false, fmt.Errorf("value of type unsupported")
+		}
+	}
+
+	switch op {
+	case datastore.OperatorEqual:
+		return val == operand, nil
+	case datastore.OperatorNotEqual:
+		return val != operand, nil
+	case datastore.OperatorGreaterThan:
+		return valNum > operandNum, nil
+	case datastore.OperatorGreaterThanOrEqual:
+		return valNum >= operandNum, nil
+	case datastore.OperatorLessThan:
+		return valNum < operandNum, nil
+	case datastore.OperatorLessThanOrEqual:
+		return valNum <= operandNum, nil
+	case datastore.OperatorIn:
+		os, err := makeSliceOfInterfaces(operand)
+		if err != nil {
+			return false, fmt.Errorf("operand error: %w", err)
+		}
+
+		for _, o := range os {
+			if o == val {
+				return true, nil
+			}
+		}
+		return false, nil
```

Comment:
nits, it would be better to make this implementation a function and return directory.

Reply:
I think it's not that different since the logic to return is different between cases 🤔